dex/src/test/java/io/crate/testing/BatchSimulatingBatchIt.java

```diff
@@ -0,0 +1,144 @@
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * An BatchIterator implementation which delegates to another one, but adds "fake" batches.
+ */
+public class BatchSimulatingBatchIt implements BatchIterator {
+
+    private final BatchIterator delegate;
+    private final int batchSize;
+    private final Executor executor;
+    private final PrimitiveIterator.OfLong loadNextDelays;
+    private final int numBatches;
+
+    private int numSuccessMoveNextCallsInBatch = 0;
+    private int currentBatch = 0;
+    private Row currentRow = OFF_ROW;
+    private final AtomicBoolean currentlyLoading = new AtomicBoolean(false);
+
+    /**
+     * @param batchSize how many {@link #moveNext()} calls are allowed per batch before it returns false
+     * @param numBatches how many {@link #loadNextBatch()} calls are allowed after {@code delegate.allLoaded()} is true.
+     *                   (This is an upper limit, if a consumer calls moveNext correctly, the actual number may be lower)
+     */
+    public BatchSimulatingBatchIt(BatchIterator delegate, int batchSize, int numBatches) {
+        assert batchSize > 0 : "batchSize must be greater than 0. It is " + batchSize;
+        assert numBatches > 0 : "numBatches must be greater than 0. It is " + numBatches;
+        this.delegate = delegate;
+        this.numBatches = numBatches;
+        this.batchSize = batchSize;
+
+        this.loadNextDelays = new Random(System.currentTimeMillis()).longs(0, 300).iterator();
+        this.executor = ForkJoinPool.commonPool();
+    }
+
+    @Override
+    public void moveToStart() {
+        if (currentlyLoading.get()) {
+            throw new IllegalStateException("Must not call moveToStart during loadNextBatch");
+        }
+        currentRow = OFF_ROW;
+        delegate.moveToStart();
+        numSuccessMoveNextCallsInBatch = 0;
+    }
+
+    @Override
+    public boolean moveNext() {
+        if (currentlyLoading.get()) {
+            throw new IllegalStateException("Must not call moveNext during loadNextBatch");
+        }
+        if (numSuccessMoveNextCallsInBatch == batchSize) {
+            currentRow = OFF_ROW;
+            return false;
+        }
+        if (delegate.moveNext()) {
+            numSuccessMoveNextCallsInBatch++;
+            currentRow = delegate.currentRow();
+            return true;
+        }
+
+        if (delegate.allLoaded()) {
+            currentBatch = numBatches;
+        }
+        currentRow = OFF_ROW;
+        return false;
+    }
+
+    @Override
+    public Row currentRow() {
+        if (currentlyLoading.get()) {
+            throw new IllegalStateException("Must not call currentRow during loadNextBatch");
+        }
+        return currentRow;
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    @Override
+    public CompletionStage<?> loadNextBatch() {
+        currentlyLoading.compareAndSet(false, true);
```

Comment:
If this returns false the client called loadNextBatch whilst the current batch is loading. Should we throw exception here?